util/src/rlp/rlpcompression.rs

```diff
@@ -0,0 +1,245 @@
+	match rlp.iter()
+	.fold((false, RlpStream::new_list(rlp.item_count())),
+		|(is_some, mut acc), subrlp| {
+  		let new = f(&subrlp);
+  		if let Some(ref insert) = new {
+  			acc.append_raw(&insert[..], 1);
+  		} else {
+  			acc.append_raw(subrlp.as_raw(), 1);
+  		}
+  		(is_some || new.is_some(), acc)
+  	  }) {
+  	(true, s) => Some(s.drain()),
+  	_ => None,
+  }
+}
+
+/// Replace common RLPs with invalid shorter ones.
+fn simple_compress(rlp: &UntrustedRlp, swapper: &InvalidRlpSwapper) -> ElasticArray1024<u8> {
+	if rlp.is_data() {
+		to_elastic(swapper.get_invalid(rlp.as_raw()).unwrap_or(rlp.as_raw()))
+	} else {
+		map_rlp(rlp, |r| Some(simple_compress(r, swapper))).unwrap_or(to_elastic(rlp.as_raw()))
+	}
+}
+
+/// Recover valid RLP from a compressed form.
+fn simple_decompress(rlp: &UntrustedRlp, swapper: &InvalidRlpSwapper) -> ElasticArray1024<u8> {
+	if rlp.is_data() {
+		to_elastic(swapper.get_valid(rlp.as_raw()).unwrap_or(rlp.as_raw()))
+	} else {
+		map_rlp(rlp, |r| Some(simple_decompress(r, swapper))).unwrap_or(to_elastic(rlp.as_raw()))
+	}
+}
+
+/// Replace common RLPs with invalid shorter ones, None if no compression achieved.
+/// Tries to compress data insides.
+fn deep_compress(rlp: &UntrustedRlp, swapper: &InvalidRlpSwapper) -> Option<ElasticArray1024<u8>> {
+	let simple_swap = ||
+		swapper.get_invalid(rlp.as_raw()).map(|b| to_elastic(&b));
+	if rlp.is_data() {
+		// Try to treat the inside as RLP.
+		return match rlp.payload_info() {
+			// Shortest decompressed account is 70, so simply try to swap the value.
+			Ok(ref p) if p.value_len < 70 => simple_swap(),
+			_ => {
+				if let Ok(d) = rlp.data() {
+					let internal_rlp = UntrustedRlp::new(&d);
+					if let Some(new_d) = deep_compress(&internal_rlp, swapper) {
+						// If compressed put in a special list, with first element being invalid code.
+						let mut rlp = RlpStream::new_list(2);
+						rlp.append_raw(&[0x81, 0x7f], 1);
+						rlp.append_raw(&new_d[..], 1);
+						return Some(rlp.drain());
+					}
+				}
+				simple_swap()
+			},
+		};
+	}
+	// Iterate through RLP while checking if it has been compressed.
+	map_rlp(rlp, |r| deep_compress(r, swapper))
+}
+
+/// Recover valid RLP from a compressed form, None if no decompression achieved.
+/// Tries to decompress compressed data insides.
+fn deep_decompress(rlp: &UntrustedRlp, swapper: &InvalidRlpSwapper) -> Option<ElasticArray1024<u8>> {
+	let simple_swap = ||
+		swapper.get_valid(rlp.as_raw()).map(|b| to_elastic(&b));
+	// Simply decompress data.
+	if rlp.is_data() { return simple_swap(); }
+	match rlp.item_count() {
+		// Look for special compressed list, which contains nested data.
+		2 if rlp.at(0).map(|r| r.as_raw() == &[0x81, 0x7f]).unwrap_or(false) =>
+			rlp.at(1).ok().map_or(simple_swap(),
+			|r| deep_decompress(&r, swapper).map(|d| { let v = d.to_vec(); encode(&v) })),
+		// Iterate through RLP while checking if it has been compressed.
+		_ => map_rlp(rlp, |r| deep_decompress(r, swapper)),
+  	}
```

Comment:
*edit, same as before. My comment is probably not valid, cause you are working on malformed data.